src/map/battleground.cpp

```diff
@@ -222,59 +621,863 @@ int bg_team_get_id(struct block_list *bl)
+		return BG_APPLY_PLAYER_LEVEL; // Level too low
+
+	if (bg->max_lvl && sd->status.base_level > bg->max_lvl)
+		return BG_APPLY_PLAYER_LEVEL; // Level too high
+
+	if (bg_queues.empty()) {
+		r = rnd() % 2 != 0;
+
+		queue = bg_queue_create(bg);
+
+		if (!r)
+			queue->teama_members.push_back(sd);
+		else
+			queue->teamb_members.push_back(sd);
+
+		sd->bg_queue = queue;
+		bg_queues.insert(bg_queues.begin(), queue);
+		return BG_APPLY_ACCEPT;
+	} else {
+		r = rnd() % 2 != 0;
+
+		for (const auto &it : bg_queues) {
+			try {
+				queue = it;
+			} catch (std::out_of_range &) {
+				continue;
+			}
+
+			if (queue->in_ready_state)
+				continue;
+			if (!r) {
+				if (queue->teama_members.size() != queue->required_players) {
+					sd->bg_queue = queue;
+					queue->teama_members.push_back(sd);
+
+					if (queue->teama_members.size() == bg->required_players && queue->teamb_members.size() == bg->required_players) // Enough players have joined
+						bg_queue_on_ready(name, queue);
+					return BG_APPLY_ACCEPT;
+				} else if (queue->teamb_members.size() != queue->required_players) {
+					sd->bg_queue = queue;
+					queue->teamb_members.push_back(sd);
+
+					if (queue->teama_members.size() == bg->required_players && queue->teamb_members.size() == bg->required_players) // Enough players have joined
+						bg_queue_on_ready(name, queue);
+					return BG_APPLY_ACCEPT;
+				}
+			} else {
+				if (queue->teamb_members.size() != queue->required_players) {
+					sd->bg_queue = queue;
+					queue->teamb_members.push_back(sd);
+
+					if (queue->teama_members.size() == bg->required_players && queue->teamb_members.size() == bg->required_players) // Enough players have joined
+						bg_queue_on_ready(name, queue);
+					return BG_APPLY_ACCEPT;
+				} else if (queue->teama_members.size() != queue->required_players) {
+					sd->bg_queue = queue;
+					queue->teama_members.push_back(sd);
+
+					if (queue->teama_members.size() == bg->required_players && queue->teamb_members.size() == bg->required_players) // Enough players have joined
+						bg_queue_on_ready(name, queue);
+					return BG_APPLY_ACCEPT;
+				}
+			}
+		}
+	}
+
+	queue = bg_queue_create(bg);
+	r = rnd() % 2 != 0;
+
+	if (!r)
+		queue->teama_members.push_back(sd);
+	else
+		queue->teamb_members.push_back(sd);
+
+	sd->bg_queue = queue;
+	bg_queues.insert(bg_queues.begin(), queue);
+	return BG_APPLY_ACCEPT;
+}
+
+/**
+ * Join a party onto the same side of a Battleground
+ * @param name: Battleground name
+ * @param sd: Player who requested to join the battlegrounds
+ * @return @see e_bg_queue_apply_ack
+ */
+enum e_bg_queue_apply_ack bg_queue_join_party(char *name, struct map_session_data *sd)
+{
+	struct party_data *p = party_search(sd->status.party_id);
+
+	if (!p)
+		return BG_APPLY_INVALID_APP; // Someone has bypassed the client check for being in a party
+
+	uint32 pl_cid = 0;
+
+	for (const auto &it : p->party.member) {
+		if (it.leader) {
```

Comment:
Checking for the party guild leader should be moved up here like so
```suggestion
		if (it.leader) {
		    if (sd->status.char_id != it.char_id) {
		        return BG_APPLY_PARTYGUILD_LEADER;
		}
```

This would eliminate the need of `pl_cid` which is only used once